docs/source/bundles.rst

```diff
@@ -0,0 +1,275 @@
+
+.. code-block:: bash
+
+   $ QUANDL_API_KEY=<api-key> python -m zipline ingest quandl
+
+though we may still run ``ingest`` as an anonymous quandl user (with no API
+key). We may also set the ``QUANDL_DOWNLOAD_ATTEMPTS`` environment variable to
+an integer which is the number of attempts that should be made to download data
+from quandls servers. By default this will be 5, meaning that we will retry each
+attempt 5 times.
+
+.. note::
+
+   This is not the total number of allowed failures, just the number of allowed
+   failures per request. This loader will make one request per 100 equities for
+   the metadata followed by one request per equity.
+
+
+Yahoo Bundle Factories
+``````````````````````
+
+Zipline also ships with a factory function for creating a data bundle out of a
+set of tickers from yahoo: :func:`~zipline.data.bundles.yahoo_equities`.
+This makes it easy to pre-download and cache the data for a set of equities from
+yahoo. This includes daily pricing data along with splits, cash dividends, and
+inferred asset metadata. To create a bundle from a set of equities, add the
+following to your ``~/.zipline/extensions.py`` file:
+
+.. code-block:: python
+
+   from zipline.bundles import register, yahoo_equities
+
+   # these are the tickers you would like data for
+   equities = {
+       'AAPL',
+       'MSFT',
+       'GOOG',
+   }
+   register(
+       'my-yahoo-equities-bundle',  # name this whatever you like
+       equities,
+   )
+
+
+This may now be used like:
+
+.. code-block:: bash
+
+   $ python -m zipline ingest my-yahoo-equities-bundle
+   $ python -m zipline run -f algo.py --bundle my-yahoo-equities-bundle
+
+
+More than one yahoo equities bundle may be registered as long as they use
+different names.
+
+Writing a New Bundle
+~~~~~~~~~~~~~~~~~~~~
+
+Data bundles exist to make it easy to use different data sources with
+zipline. To add a new bundle, one must implement an ingest function.
+
+This function is responsible for loading the data into memory and passing it to
+a set of writer objects provided by zipline to convert the data to zipline's
+internal format. The ingest function may work by downloading data from a remote
+location like the ``quandl`` bundle or yahoo bundles or it may just load files
+that are already on the machine. The function is provided with writers that will
+write the data to the correct location transactionally. If an ingestion fails
+part way through the bundle will not be written in an incomplete state.
+
+The signature of the ingest function should be:
+
+.. code-block:: python
+
+   ingest(environ,
+          asset_db_writer,
+          minute_bar_writer,
+          daily_bar_writer,
+          adjustment_writer,
+          calendar,
+          cache,
+          show_progress)
+
+``environ``
+```````````
+
+``environ`` is a mapping representing the environment variables to use. This is
+where any custom arguments needed for the ingestion should be passed, for
```

Comment:
This is where the bundle author should find any arguments that it needs. For example, quandl needs to know what the auth key is so it uses the env. Do you have a suggestion on how to word this?